File: leetcode/medium/1249-min-remove/a.cpp
/**
 * @file a.cpp
 * @author Vortexx2
 * @brief Problem 1249 - Minimum Remove to Make Valid Parentheses
 * @approach Scanning string twice. Once forward and then backwards.
 * @date 05-07-2021
 * 
 * TLE. Don't know why. O(n).
 */
#include <iostream>
#include <stack>
#include <string>
#include <vector>

using namespace std;

class Solution
{
public:
  string minRemoveToMakeValid(string s)
  {
    if (s == "")
      return "";

    int count = 0;
    string temp = "";

    for (auto &ch : s)
    {
      if (ch == '(')
      {
        count++;
      }
      else if (ch == ')')
      {
        if (!count)
          continue;

        count--;
      }
      temp += ch;
    }

    count = 0;
    string res = "";

    for (int j = temp.size() - 1; j >= 0; j--)
    {

      if (temp[j] == ')')
      {
        count++;
      }

      else if (temp[j] == '(')
      {
        if (!count)
          continue;

        count--;
      }

      res = temp[j] + res;
    }

    return res;
  }
};

File: leetcode/medium/1249-min-remove/a.py
"""
  @author Vortexx2
  Problem 1249 - Minimum Remove to Make Valid Parentheses
  @approach Scanning string twice. Once forward and then backwards.

  Runtime - 432 ms
  Memory Usage - 15.9 MB
"""
class Solution:
  def minRemoveToMakeValid(self, s: str) -> str:

    count = 0
    temp = ""

    for ch in s:

      if ch == '(':
        count += 1

      elif ch == ')':
        if (not count):
          continue

        count -= 1

      temp += ch

    count = 0
    res = ""
    for ch in reversed(temp):

      if ch == ')':
        count += 1

      elif ch == '(':

        if not count:
          continue

        count -= 1

      res = ch + res

    return res
